src/components/Provider.tsx

```diff
@@ -4,32 +4,40 @@ import { createSubscription } from '../utils/Subscription'
 import { useIsomorphicLayoutEffect } from '../utils/useIsomorphicLayoutEffect'
 import { Action, AnyAction, Store } from 'redux'
 
-export interface ProviderProps<A extends Action = AnyAction> {
+export interface ProviderProps<A extends Action = AnyAction, S = any> {
   /**
    * The single Redux store in your application.
    */
-  store: Store<any, A>
+  store: Store<S, A>
+
+  /**
+   * An optional server state snapshot. Will be used during initial hydration render if available, to ensure that the UI output is consistent with the HTML generated on the server.
+   */
+  serverState?: S
+
   /**
    * Optional context to be used internally in react-redux. Use React.createContext() to create a context to be used.
    * If this is used, you'll need to customize `connect` by supplying the same context provided to the Provider.
    * Initial value doesn't matter, as it is overwritten with the internal state of Provider.
    */
-  context?: Context<ReactReduxContextValue<any, A>>
+  context?: Context<ReactReduxContextValue<S, A>>
   children: ReactNode
 }
 
 function Provider<A extends Action = AnyAction>({
   store,
   context,
   children,
+  serverState,
 }: ProviderProps<A>) {
   const contextValue = useMemo(() => {
     const subscription = createSubscription(store)
```

Comment:
q: this component still relies on a custom Subscription and custom notification propagation etc. Doesn't `useSyncExternalStore` solve the ordering issue?

Reply:
Nope, not at all. `uSES` still creates store subscriptions same as always, and React still runs effects bottom-to-top on mount.  So, children still subscribe before parents, etc, and we still have to do work on our side to enforce the "top-down updates" requirement for `connect`.

The only thing in React that has ever helped with that was when we switched v6 to read state updates from context. It _worked_.... which is why it was so frustrating that it also turned out to be the worst case for performance updates.

Reply:
> we still have to do work on our side to enforce the "top-down updates" requirement for connect.

Isn't the problem relevant to `useSelector` too? Is React subscribing to the "external store" (Redux) with `uSES` still in the order that makes shenanigans required? Is the zombie children problem still relevant with `uSES`? Or is the problem somewhat fixed with automatic batching and top-down rerender that should happen with it? As automatic batching should end up rerendering things in top-down order, right?

Reply:
Yes, the problem _is_ entirely relevant to `useSelector`, and there's absolutely nothing we can do to fix it - because hooks can't render context providers, so we can't enforce top-down behavior.  That's why I documented all the "zombie child" concepts in the hooks API page in the first place:

https://react-redux.js.org/api/hooks#stale-props-and-zombie-children

And no, batching still has nothing to do with this problem, because it's _the subscription callbacks themselves_ that are the problem, and those run _outside_ of React when the store is updated.

The hacky-but-valid workaround that both `useSelector` and `uSES` take is catching errors in a selector when subscribers are notified, forcing a re-render anyway, re-running the selector _during_ the render, and seeing if it explodes again or succeeds.

Reply:
Ok, I understand how automatic batching is not the remedy for the problem.

> That workaround is good enough that it does actually mostly mitigate the issue. But the root cause of the problem still exists - components can end up subscribing to the store in an order where a child is notified before its parent, and if the child depends on using props in its selector, those props can be stale or the child may be reading state right before its parent is about to stop rendering the child.

I must say that I'm somewhat disappointed with the `uSES` API then. I was under the impression that this problem would actually be solved by this API. This problem is affecting all external stores and thus the implementation that would guard users against this problem is still very complicated - so marketing `uSES` as a tool solving external store problems is somewhat misleading. I think that what it solves was way easier to solve in the userland (maybe apart from allowing inline selectors, as that was super tricky) than the zombie children problem - solving this problem at the React level would be the bigger "win" for the library authors.

@acdlite is solving this problem out of the scope of `uSES`? How are other external stores supposed to handle those situations? Should we use the hacky solution outlined above by @markerikson?

Reply:
FWIW `uSES` already implements that "hacky solution" :) and specifically inspired by `useSelector` doing it first.

Really the problem is a mismatch in data shapes: a linear array of subscriber callbacks executed in order, vs a tree of components where A) mount effects are executed bottom-to-top, and B) components can be added and removed over time.

Reply:
> If there were some way to actually correlate "adding a subscriber" with "actual position in the React component tree", then it would be theoretically possible to build a smarter notification system that actually notifies subscribers in tree parent->child order. But there's no way to know where a given component is in the tree.

I would put the asterisk here that it's no way to know that in the userland - that's why I think this is the problem worth solving in the React itself (in the `uSES`).

> FWIW uSES already implements that "hacky solution" :) and specifically inspired by useSelector doing it first.

Hm, so I'm a little bit confused right now. I wonder if I understand the whole zombie children problem right now - I think I understand the problem but I might not understand how the problem is solved in React Redux today.

So to get us on the same page. This is what I think right now:
1. zombie children problem is caused~ by subscriber callbacks being called "out of order" because a child's callback can be called **before** the parent's callback - this leads to errors being thrown when items in the store are deleted during the latest dispatch~
2. the error is captured by React Redux and a new rerender is being scheduled
3. that new rerender is already executed when props are already updated and thus the selector should no longer throw or a rerender for that child might not happen at all because that component could have been unmounted by the render that has thrown (unmounted by its parent rerendering with the selected state that already doesn't contain the deleted item)
4. so in a sense - the whole zombie children problem is already "solved" by capturing errors and rerendering, the point is that users of React Redux don't have to deal with this. They might occasionally observe weird-ish things when debugging, but that's acceptable.
5. however, there is still the cost of the child being rendered at all as that could have been prevented if subscriptions would merely mark components as "dirty" and the state would actually be selected in render - with such a solution the child's subscriber wouldn't select anything that could throw (if only this would be batched and performed in the top-down order). Or if the item wasn't completely removed and the error is caused by partial lack of information etc then we get 2 renders there instead of one.

Is this reasoning correct? I think that I see where the `use-sync-external-store/shim` implements the try/catch dance and that looks good. It means that libraries using that won't have to deal with this problem at all because that's covered there. I wonder though - if the implementation of the "native" `useSyncExternalStore` the same or is the issue mitigated there because things can actually be executed in the "correct" order in React 18?

Reply:
Yep, that's all correct.  (fwiw I don't think the "cost" thing in step 5 is really a meaningful issue.)

You can see that the actual `useSyncExternalStore` does the same 'rerender if error" behavior here:

https://github.com/facebook/react/blob/3dc41d8a2590768a6ac906cd1f4c11ca00417eee/packages/react-reconciler/src/ReactFiberHooks.new.js#L1482-L1491

Reply:
> Yep, that's all correct. (fwiw I don't think the "cost" thing in step 5 is really a meaningful issue.)

I agree that this could have been somewhat exaggerated on my side. If this is only affecting subscribers and not rerenders then I guess this is somewhat OK (I assume that the inner component might in fact not even rerender, even if it throws, if it gets "unmounted" by the parent). However, I still find it a little bit strange that React 18 doesn't try to mitigate this all together in a smarter way - but dunno, maybe the smarter way would have some perf implications for the whole thing.